app/forms/edit_appointment_form.rb

```diff
@@ -0,0 +1,64 @@
+class EditAppointmentForm
+  include ActiveModel::Model
+
+  ATTRIBUTES = %i(
+    id
+    reference
+    booking_location
+    booking_request
+  ).freeze
+
+  EDITABLE_ATTRIBUTES = %i(
+    name
+    email
+    phone
+    location_id
+    guider_id
+  ).freeze
+
+  attr_writer(*EDITABLE_ATTRIBUTES)
+
+  delegate(*ATTRIBUTES, to: :location_aware_appointment)
+
+  delegate :guiders, to: :booking_location
+
+  delegate :primary_slot, :secondary_slot, :tertiary_slot, to: :booking_request
+
+  def initialize(location_aware_appointment, params)
+    @location_aware_appointment = location_aware_appointment
+
+    normalise_time(params)
+    super(params)
+  end
+
+  def proceeded_at
+    @proceeded_at ||= location_aware_appointment.proceeded_at
+
+    Time.zone.parse(@proceeded_at.to_s)
+  end
+
+  def flattened_locations
+    FlattenedLocationMapper.map(booking_location)
+  end
+
+  # memoized readers to use `params` when present
+  EDITABLE_ATTRIBUTES.each do |attribute|
```

Comment:
I'm not sure - given the small set of `EDITABLE_ATTRIBUTES` whether just writing these out in regular code is just a simpler approach. Personally, I have trouble with metaprogramming when it's contained.